src/dataframerow/utils.jl

```diff
@@ -167,43 +185,83 @@ function row_group_slots(cols::NTuple{N,<:Union{CategoricalVector,PooledVector}}
     # but it needs to remain reasonable compared with the size of the data frame.
     if prod(Int128.(ngroupstup)) > typemax(Int) || ngroups > 2 * length(groups)
         return invoke(row_group_slots,
-                      Tuple{Tuple{Vararg{AbstractVector}}, Val,
-                            Union{Vector{Int}, Nothing}, Bool},
-                      cols, hash, groups, skipmissing)
+                      Tuple{Tuple{Vararg{AbstractVector}}, Any, Val,
+                            Union{Vector{Int}, Nothing}, Bool, Bool},
+                      cols, refpools, hash, groups, skipmissing, sort)
     end
 
     seen = fill(false, ngroups)
-    # Compute vector mapping missing to -1 if skipmissing=true
-    refmaps = map(cols) do col
-        nlevs = nlevels(col)
-        refmap = collect(-1:(nlevs-1))
-        # First value in refmap is only used by CategoricalArray
-        # (corresponds to ref 0, i.e. missing values)
-        refmap[1] = skipmissing ? -1 : nlevs
-        if col isa PooledArray{>: Missing} && skipmissing
-            missingind = get(col.invpool, missing, 0)
-            if missingind > 0
-                refmap[missingind+1] = -1
-                refmap[missingind+2:end] .-= 1
+    refs = map(DataAPI.refarray, cols)
+    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
+    firstinds = map(firstindex, refpools)
+    # TODO: when skipmissing=true, do not include missing values
+    # when checking whether pool is sorted
+    if sort && !all(issorted, refpools)
+        # Compute vector mapping missing to -1 if skipmissing=true
+        refmaps = map(cols, refpools) do col, refpool
+            refmap = collect(0:length(refpool)-1)
+            if skipmissing
+                fi = firstindex(refpool)
+                missingind = findfirst(ismissing, refpool)
+                if missingind !== nothing
+                    mi = something(missingind)
+                    refmap[mi-fi+1] = -1
+                    refmap[mi-fi+2:end] .-= 1
+                end
+                if sort
+                    nm = missingind === nothing ? eachindex(refpool) :
+                        setdiff(eachindex(refpool), missingind)
+                    perm = sortperm(view(refpool, nm))
+                    invpermute!(view(refmap, nm .- fi .+ 1), perm)
+                end
+            elseif sort
+                # FIXME: collect is needed for CategoricalRefPool
+                perm = sortperm(collect(refpool))
```

Comment:
Without this I get:
```julia
ERROR: MethodError: no method matching similar(::Type{Array{Int64,1}}, ::Tuple{UnitRange{Int64}})
```

I'm afraid it could be hard to fix in CategoricalArrays, since we would need to return an `OffsetArray`. Also since our `refmap` uses 1-based indexing, we actually don't want to preserve custom indices. Calling `collect` is OK in terms of API but it's a bit of a waste in terms of copying...

Reply:
I think it is not a problem as it should be very fast - right?

Reply:
Yeah it's not too bad. If there are many levels it's a bit silly to make a copy of them, but not a big deal. I've removed the FIXME.